patches/server/0010-Adventure.patch

```diff
@@ -1213,6 +1039,72 @@ index 0000000000000000000000000000000000000000..d496a9a6ad229e42f1c44e31eafa6974
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/adventure/ImprovedChatDecorator.java
+@@ -0,0 +1,60 @@
++package io.papermc.paper.adventure;
++
++import io.papermc.paper.event.player.AsyncChatCommandDecorateEvent;
++import io.papermc.paper.event.player.AsyncChatDecorateEvent;
++import net.minecraft.commands.CommandSourceStack;
++import net.minecraft.network.chat.ChatDecorator;
++import net.minecraft.server.MinecraftServer;
++import net.minecraft.network.chat.Component;
++import net.minecraft.server.level.ServerPlayer;
++import org.bukkit.craftbukkit.entity.CraftPlayer;
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.framework.qual.DefaultQualifier;
++
++import javax.annotation.Nullable;
++import java.util.concurrent.CompletableFuture;
++
++@DefaultQualifier(NonNull.class)
++public final class ImprovedChatDecorator implements ChatDecorator {
++    private final MinecraftServer server;
++
++    public ImprovedChatDecorator(MinecraftServer server) {
++        this.server = server;
++    }
++
++    @Override
++    public CompletableFuture<Component> decorate(@Nullable ServerPlayer sender, Component message) {
++        return decorate(server, sender, null, message);
++    }
++
++    @Override
++    public CompletableFuture<Component> decorate(@Nullable ServerPlayer sender, @Nullable CommandSourceStack commandSourceStack, Component message) {
++        return decorate(server, sender, commandSourceStack, message);
++    }
++
++    private static CompletableFuture<Component> decorate(final MinecraftServer server, final @Nullable ServerPlayer player, final @Nullable CommandSourceStack commandSourceStack, final Component originalMessage) {
++        return CompletableFuture.supplyAsync(() -> {
++            // Make a copy so that if the event if cancelled we can guarantee the original is being returned.
++            final net.kyori.adventure.text.Component originalMessageCopy = PaperAdventure.asAdventure(originalMessage.copy());
++            final net.kyori.adventure.text.Component initialResult = PaperAdventure.asAdventure(originalMessage.copy());
++
++            final @Nullable CraftPlayer craftPlayer = player == null ? null : player.getBukkitEntity();
++
++            final AsyncChatDecorateEvent event;
++            if (commandSourceStack != null) {
++                // TODO more command decorate context
++                event = new AsyncChatCommandDecorateEvent(true, craftPlayer, originalMessageCopy, initialResult);
++            } else {
++                event = new AsyncChatDecorateEvent(true, craftPlayer, originalMessageCopy, initialResult);
++            }
++
++            server.server.getPluginManager().callEvent(event);
++
++            if (!event.isCancelled() && !event.result().equals(PaperAdventure.asAdventure(originalMessage))) {
++                return PaperAdventure.asVanilla(event.result());
++            }
++
++            return originalMessage;
```

Comment:
I think this can be simplified to be
If the event is cancelled, return `originalMessage`, otherwise, return `PaperAdventure.asVanilla(event.result())`. No need for this extra equals check.